# Data Cleaning — Online Retail Dataset

## Objective

The purpose of this notebook is to inspect the cleaned Online Retail dataset and verify that the data-cleaning pipeline produced an analysis-ready dataset.

The raw dataset contains transaction-level records including invoice information, products, quantities, prices, customers, dates, and countries.

The cleaning process addresses:

- Duplicate transactions
- Cancelled invoices
- Non-positive quantities
- Non-positive unit prices
- Missing product descriptions
- Invalid invoice dates

Transactions with missing Customer IDs are retained because they can still contribute to overall sales, product, country, and time-based analyses. Customer-specific analyses will use only transactions with valid Customer IDs.

In [15]:
import pandas as pd

df = pd.read_csv(
    "../data/processed/cleaned_sales_data.csv",
    parse_dates=["invoicedate"]
)

print("Dataset loaded successfully.")

C:\Users\USer\AppData\Local\Temp\ipykernel_2884\531033123.py:3: DtypeWarning: Columns (0: invoiceno) have mixed types. Specify dtype option on import or set low_memory=False.
  df = pd.read_csv(


Dataset loaded successfully.


In [16]:
print(f"Rows: {df.shape[0]:,}")
print(f"Columns: {df.shape[1]}")

Rows: 524,878
Columns: 13


In [17]:
df.head()

,invoiceno,stockcode,description,quantity,invoicedate,unitprice,customerid,country,sales,year,month,month_name,day_of_week
0,536365,85123A,WHITE HANGING HEART T-LIGHT HOLDER,6,2010-12-01 08:26:00,2.55,17850.0,United Kingdom,15.30,2010,12,December,Wednesday
1,536365,71053,WHITE METAL LANTERN,6,2010-12-01 08:26:00,3.39,17850.0,United Kingdom,20.34,2010,12,December,Wednesday
2,536365,84406B,CREAM CUPID HEARTS COAT HANGER,8,2010-12-01 08:26:00,2.75,17850.0,United Kingdom,22.00,2010,12,December,Wednesday
3,536365,84029G,KNITTED UNION FLAG HOT WATER BOTTLE,6,2010-12-01 08:26:00,3.39,17850.0,United Kingdom,20.34,2010,12,December,Wednesday
4,536365,84029E,RED WOOLLY HOTTIE WHITE HEART.,6,2010-12-01 08:26:00,3.39,17850.0,United Kingdom,20.34,2010,12,December,Wednesday


In [18]:
df.info()

<class 'pandas.DataFrame'>
RangeIndex: 524878 entries, 0 to 524877
Data columns (total 13 columns):
 #   Column       Non-Null Count   Dtype         
---  ------       --------------   -----         
 0   invoiceno    524878 non-null  object        
 1   stockcode    524878 non-null  str           
 2   description  524878 non-null  str           
 3   quantity     524878 non-null  int64         
 4   invoicedate  524878 non-null  datetime64[us]
 5   unitprice    524878 non-null  float64       
 6   customerid   392692 non-null  float64       
 7   country      524878 non-null  str           
 8   sales        524878 non-null  float64       
 9   year         524878 non-null  int64         
 10  month        524878 non-null  int64         
 11  month_name   524878 non-null  str           
 12  day_of_week  524878 non-null  str           
dtypes: datetime64[us](1), float64(3), int64(3), object(1), str(5)
memory usage: 52.1+ MB


In [19]:
df.isnull().sum()

invoiceno           0
stockcode           0
description         0
quantity            0
invoicedate         0
unitprice           0
customerid     132186
country             0
sales               0
year                0
month               0
month_name          0
day_of_week         0
dtype: int64

In [20]:
print("Duplicate rows:", df.duplicated().sum())

print(
    "Cancelled invoices:",
    df["invoiceno"]
    .astype(str)
    .str.startswith("C")
    .sum()
)

print(
    "Non-positive quantities:",
    (df["quantity"] <= 0).sum()
)

print(
    "Non-positive prices:",
    (df["unitprice"] <= 0).sum()
)

Duplicate rows: 0
Cancelled invoices: 0
Non-positive quantities: 0
Non-positive prices: 0


## Cleaning Summary

The raw dataset contained **541,909 rows and 8 columns**.

After cleaning, the analysis-ready dataset contains **524,878 rows and 13 columns**.

A total of **17,031 rows** were removed during the cleaning process, representing approximately **3.14% of the original dataset**.

### Cleaning Decisions

The following steps were applied:

1. Removed exact duplicate transactions.
2. Converted invoice dates to a valid datetime format.
3. Removed cancelled invoices.
4. Removed transactions with non-positive quantities.
5. Removed transactions with zero or negative unit prices.
6. Removed records without product descriptions.
7. Retained transactions with missing Customer IDs for non-customer-specific analysis.
8. Created a `sales` variable using quantity × unit price.
9. Created year, month, month name, and day-of-week variables for time-based analysis.

The resulting dataset is suitable for exploratory sales analysis while preserving anonymous transactions that still contain useful business information.

In [21]:
raw_rows = 541909
clean_rows = len(df)

removed_rows = raw_rows - clean_rows
removed_percentage = (removed_rows / raw_rows) * 100

print(f"Raw rows: {raw_rows:,}")
print(f"Clean rows: {clean_rows:,}")
print(f"Rows removed: {removed_rows:,}")
print(f"Percentage removed: {removed_percentage:.2f}%")

Raw rows: 541,909
Clean rows: 524,878
Rows removed: 17,031
Percentage removed: 3.14%
